src/main/java/org/zalando/nakadi/service/kpi/publisher/EventsProcessor.java

```diff
@@ -0,0 +1,119 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import org.zalando.nakadi.service.EventPublisher;
+import org.zalando.nakadi.util.FeatureToggleService;
+import org.zalando.nakadi.util.UUIDGenerator;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+@Component
+public final class EventsProcessor {
+
+    private final static Logger LOG = LoggerFactory.getLogger(EventsProcessor.class);
+    private final EventPublisher eventPublisher;
+    private final Map<String, BlockingQueue<JSONObject>> eventTypeEvents;
+    private final ExecutorService executorService;
+    private final UUIDGenerator uuidGenerator;
+    private final FeatureToggleService featureToggleService;
+
+    private final long batchCollectionTimeout;
+    private final int batchSize;
+    private final long pollTimeout;
+    private final int eventsQueueSize;
+
+    @Autowired
+    public EventsProcessor(final EventPublisher eventPublisher,
+                           final UUIDGenerator uuidGenerator,
+                           final FeatureToggleService featureToggleService,
+                           @Value("${nakadi.kpi.batch-collection-timeout:1000}") final long batchCollectionTimeout,
+                           @Value("${nakadi.kpi.batch-size:3}") final int batchSize,
+                           @Value("${nakadi.kpi.workers:1}") final int workers,
+                           @Value("${nakadi.kpi.poll-timeout:100}") final long pollTimeout,
+                           @Value("${nakadi.kpi.events-queue-size:100}") final int eventsQueueSize) {
+        this.eventPublisher = eventPublisher;
+        this.uuidGenerator = uuidGenerator;
+        this.featureToggleService = featureToggleService;
+        this.batchCollectionTimeout = batchCollectionTimeout;
+        this.batchSize = batchSize;
+        this.pollTimeout = pollTimeout;
+        this.eventsQueueSize = eventsQueueSize;
+        this.eventTypeEvents = new ConcurrentHashMap<>();
+        this.executorService = Executors.newFixedThreadPool(workers);
+    }
+
+    private void sendEventBatch(final String etName) {
+        LOG.trace("Collecting batch for {} on {}", etName, Thread.currentThread().getName());
+        try {
+            final BlockingQueue<JSONObject> events = eventTypeEvents.get(etName);
+            if (events == null) {
+                return;
+            }
+
+            final long finishAt = System.currentTimeMillis() + batchCollectionTimeout;
+            int eventsCount = 0;
+            JSONArray jsonArray = null;
+            while (eventsCount != batchSize && System.currentTimeMillis() < finishAt) {
```

Comment:
I'm not sure if we need batching here since batching is already done by the KafkaProducer.

I would even say, let's create a different Kafka producer, not to use the same producer as user events, and just fire and forget events on it. I mean, we don't need to use the "sync" publisher for this feature.

We could remove all the async code from here and just queue the events in the the KafkaProducer directly since it's already async.

Reply:
Or if we are using the same producer, I would consider using it on a lower level, not to go through `EventPublisher`. Or is there a particular reason why you chose to use the `EventPublisher` with all of its features: validation, enrichment, partitioning and so on?

Reply:
Ah ok, timelines... hunnn I would consider adding a new method to the existing publisher to allow for async post in order to reuse all the queue and async done by the KafkaProducer not to have another buffer zone here.

Reply:
I understand your point, in the draft version I even tried to implement a separate method, which helps to avoid validation step, but there are still other steps, which taking most of the time.
The purpose of the buffers here is to avoid spending time on the client thread. The time spent is not only about publishing to Kafka, but also about timelines, validation(?), enrichment, authz, short circuiting.

Reply:
Ok, understood. If we did not go though unnecessary steps like validation, enrichment and so on, we would be able to reduce the time spent on user thread considerably, since it would be just a matter of queueing the event in the kafka producer.

But I think your solution is good and I'm not argue any further.